### ISSUE TITLE
Bug: Application memory is not reduce after using dispose function

### ISSUE THREAD

**Issue by sang-hv** [amcharts/amcharts5#756]
**Bug description**
Dear Amchart support team,

I'm a free Amchart user. Currently, I caught a bug in the library and don't know how to handle it. I will explain it down here and hopefully will get support back from you.

Environment: NuxtJS 2.15.8, Amchart: 5
Bug: Application memory is not reduce after using dispose function

Bug reproduction:
- Initialize chart in mounted:
```
mounted () {
 const { am5, am5xy, am5themesResponsive, kelly } = this.$am5()
 this.root = am5.Root.new(this.$refs.chartdiv)
 // this.root.autoResize = false

 this.root.setThemes([am5themesResponsive.new(this.root), kelly.new(this.root)])
 this.chart = this.root.container.children.push(
   am5xy.XYChart.new(this.root, {
     panY: true,
     panX: true,
     wheelY: 'zoomX'
   })
 )

 this.chart.paddingRight = 20

 this.scrollbarX = am5xy.XYChartScrollbar.new(this.root, {
   orientation: 'horizontal',
   height: 50
 })

 this.chart.set('scrollbarX', this.scrollbarX)

 this.yAxis = this.chart.yAxes.push(
   am5xy.ValueAxis.new(this.root, {
     renderer: am5xy.AxisRendererY.new(this.root, {})
   })
 )
 this.xAxis = this.chart.xAxes.push(
   am5xy.DateAxis.new(this.root, {
     renderer: am5xy.AxisRendererX.new(this.root, {}),
     baseInterval: {
       timeUnit: 'minute',
       count: 1
     }
     // groupData: true
   })
 )

 this.sbxAxis = this.scrollbarX.chart.xAxes.push(
   am5xy.DateAxis.new(this.root, {
     groupIntervals: [{ timeUnit: 'minute', count: 1 }],
     baseInterval: { timeUnit: 'minute', count: 1 },
     renderer: am5xy.AxisRendererX.new(this.root, {
       opposite: false,
       strokeOpacity: 0
     })
   })
 )

 this.sbyAxis = this.scrollbarX.chart.yAxes.push(
   am5xy.ValueAxis.new(this.root, {
     renderer: am5xy.AxisRendererY.new(this.root, {})
   })
 )
 this.chart.events.on('wheelended', this.chartRangeChanged)
 this.chart.events.on('panended', this.chartRangeChanged)

 this.scrollbarX.events.on('rangechanged', this.scrollBarRangeChanged)
}
```
- When the radio is changed, the chart will be changed. Before changing the graph, do dispose() of the current series:

```
// clear series and sbseries
if (this.chart.series.length > 0) {
  while (this.chart.series.length > 0) {
    this.chart.series.removeIndex(0).dispose()
  }
}

if (this.scrollbarX.chart.series.values.length > 0) {
  while (this.scrollbarX.chart.series.values.length > 0) {
    this.scrollbarX.chart.series.removeIndex(0).dispose()
  }
}
```
- Create series and subseries, then set data:

```
// create series
const series = this.createSeries(valueList.name, am5.color('#0000FF'), false, false, true)
series.data.setAll(this.anomalyData)
series.appear()

// create sbSeries
const sbseries = this.createSbseries()
sbseries.data.setAll(data)
sbseries.appear()
```
- Create new series and subseries function

```
createSeries (name, color, strokeDash = false, connect = false, withBullets = false) {
  const { am5, am5xy } = this.$am5()

  const series = this.chart.series.push(
    am5xy.LineSeries.new(this.root, {
      name,
      xAxis: this.xAxis,
      yAxis: this.yAxis,
      valueYField: 'value',
      valueXField: 'date',
      stroke: color,
      fill: color,
      connect,
      legendLabelText: '{name}',
      legendRangeLabelText: '{name}'
      // minDistance: 20
    })
  )

  series.strokes.template.setAll({
    strokeWidth: 1
  })

  if (strokeDash) {
    series.strokes.template.setAll({
      strokeDasharray: [10, 5]
    })
  }

  series.data.processor = am5.DataProcessor.new(this.root, {
    dateFields: ['date']
  })

  if (withBullets) {
    series.bullets.push(function (root) {
      return am5.Bullet.new(root, {
        sprite: am5.Circle.new(root, {
          radius: 2,
          fill: series.get('fill')
        })
      })
    })
  }

  return series
},
createSbseries () {
  const { am5, am5xy } = this.$am5()

  const sbseries = this.scrollbarX.chart.series.push(
    am5xy.LineSeries.new(this.root, {
      xAxis: this.sbxAxis,
      yAxis: this.sbyAxis,
      valueYField: 'value',
      valueXField: 'date',
      connect: false
      // minDistance: 20
    })
  )
  sbseries.data.processor = am5.DataProcessor.new(this.root, {
    dateFields: ['date']
  })

  return sbseries
}
```
=> The problem I have right now is when I call the function this.chart.dispose() or this.root.dispose(), the application is not released.

**Comment by sassomedia** [comment]
> When the radio is changed, the chart will be changed. Before changing the graph, do dispose() of the current series

How about redrawing the whole chart when the radio button changes? Instead of disposing of the series, dispose the chart (`this.root.dispose()`) and create a new chart with new series settings. Would that not work?

**Comment by sang-hv** [comment]
> > When the radio is changed, the chart will be changed. Before changing the graph, do dispose() of the current series
> 
> How about redrawing the whole chart when the radio button changes? Instead of disposing of the series, dispose the chart (`this.root.dispose()`) and create a new chart with new series settings. Would that not work?

i did but it not work!

**Comment by martynasma** [comment]
Does the memory consumption grow to the point where performance of the browser starts to degrade or becomes unresponsive?

Memory in browser's are being released by a garbage collection process, which intelligently checks available memory and does not clear up unused variables just because they are not used anymore. They follow own fuzzy logic about when and what to clean up.

**Comment by sang-hv** [comment]
> Does the memory consumption grow to the point where performance of the browser starts to degrade or becomes unresponsive?
> 
> Memory in browser's are being released by a garbage collection process, which intelligently checks available memory and does not clear up unused variables just because they are not used anymore. They follow own fuzzy logic about when and what to clean up.

memory consumption gradually increases, so large that the browser crashes OutOfMemory

**Comment by martynasma** [comment]
Gotcha. Can you doublecheck if you're running the latest version (5.2.47)?

**Comment by martynasma** [comment]
Would you be able to prepare a runnable demo which we can test? E.g. on CodeSandbox or similar.

**Comment by whitebeard28** [comment]
https://codesandbox.io/p/sandbox/amcharts-memory-test-861zbw?file=%2FREADME.md
We have made a demo version in here. You can test it. In real product, the data can be more than this. Around 4500 points.

**Comment by zeroin** [comment]
I've tried your script and I must say I don't see a memory leak, the memory is cleared fine, attached is an image. However your chart is really heavy and I believe that on some slower computers browser might crash because of overload. So my suggestion would be to optimize chart first. setting `minBulletDistance` to some number like 20 would make huge difference, as you would only see bullets when zoomed-in and it's the amount of bullets (5000) which makes the biggest problem. I would also say that having all 5 series in scrollbar doesn't make a lot of sense as it's simply not readable.

**Comment by zeroin** [comment]
<img width="1042" alt="image (3)" src="https://user-images.githubusercontent.com/5336255/212839524-01dcc652-5a25-4de9-87af-18248f088cd5.png">

**Comment by whitebeard28** [comment]
[https://drive.google.com/file/d/1rSHJDQRKwoL9dsDxk8LMRkMkzfxLOFKV/view?usp=sharing](Ram crash video)
We knew about the minBulletDistance function, we may apply that change to our code too. However, due to the need of changing between graphs a lot, every time changing the graph, we call to the dispose function to release the memory but it didn't work. You can see it in the video above.

**Comment by zeroin** [comment]
I see, strange that I don't see this leak when inspecting performance. Can I ask you to prepare the same demo outside of nuxt, a working codepen would be best. This way I could do testing on our dev version, also it would be clear that this is not some 3rd party related problem.

**Comment by adiospl** [comment]
I found two major problems in the code you delivered.

First of all - keep away anything related to amCharts from Vue reactivity system. If anything has an observer Garbage Collector will not be able to clear memory. Besides, Vue will try to observe the whole, giant object this.root, this.chart etc.

Secondly - aways prefer templated over directly passing properties during initialization.

`
      const bulletTemplate = am5.Template.new({
        radius: 2,
        fill: series.get('fill'),
      })

      if (withBullets) {
        series.bullets.push(function (root) {
          const sprite = am5.Circle.new(root, {}, bulletTemplate)
          return am5.Bullet.new(root, { sprite })
        })
      }`
      
     You can compare performance gains with [forked version](https://codesandbox.io/p/sandbox/amcharts-memory-test-forked-5oix6w?file=%2Fpages%2Findex.vue&selection=%5B%7B%22endColumn%22%3A8%2C%22endLineNumber%22%3A252%2C%22startColumn%22%3A1%2C%22startLineNumber%22%3A241%7D%5D).
     
And a request to the authors of the library. Please, create a documentation page with a detailed description of how to write efficient code. It is difficult to write good code if even the demo includes performance anti-patterns....

**Comment by Pauan** [comment]
@sang-hv You must never put amCharts objects inside of `data`, because Vue does all sorts of weird things with reactive objects. This is [mentioned in the documentation](https://www.amcharts.com/docs/v5/getting-started/integrations/vue/#Important_note).

Also, in `beforeDestroy` you are disposing the chart, but that's wrong, you need to dispose the root. This is also mentioned [in our documentation](https://www.amcharts.com/docs/v5/getting-started/root-element/#disposing).

I also improved your `updateLegend` method.

Here is an updated codesandbox which fixes those problems:

https://codesandbox.io/p/sandbox/amcharts-memory-test-forked-6qpwgy

**Comment by Pauan** [comment]
> @adiospl And a request to the authors of the library. Please, create a documentation page with a detailed description of how to write efficient code. It is difficult to write good code if even the demo includes performance anti-patterns....

Could you show which demo contains performance anti-patterns? Our examples are usually written with good code, so there shouldn't be any problem following them.

**Comment by Pauan** [comment]
> @adiospl aways prefer templated over directly passing properties during initialization.

Templates are slower than direct properties, but they're very useful when instantiating many objects with the same properties (like bullets).

**Comment by shohei-hacarus** [comment]
Thanks to your advice, the problem in our app @sang-hv mentioned has been resolved.

I have made the following modifications;

- Remove away all amCharts objects from Vue's data property
- Execute `this.root.dispose()` and initialize the graph each time switching the data to be displayed

Memory release is now working properly and the speed of graph display, scrolling and zoom has been also improved!

Your documentation and sample codes are already helpful enough to solve our problem.
Thank you again for your help.
